### PR TITLE
[metal] Add mem_offset_in_parent to AOT module

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #

This PR stores the memory offset of each field within the root buffer.

### CODE REVIEW

File: taichi/backends/wasm/aot_module_builder_impl.h [hunk 1]
```diff
@@ -24,7 +24,8 @@ class AotModuleBuilderImpl : public AotModuleBuilder {
   void add_per_backend_tmpl(const std::string &identifier,
                             const std::string &key,
                             Kernel *kernel) override;
-  void add_per_backend_field(const std::string &Identifier,
+  void add_field_per_backend(const std::string &Identifier,
```
Comment (bobcao3) [inline taichi/backends/wasm/aot_module_builder_impl.h:27]: When we have the snode, is the identifier still important?

**Review by bobcao3 — APPROVED** [review]
other than that LGTM!

**Review by ailzhang — APPROVED** [review]
Looks great! 
(now I know what you meant about missing mem_offset in the opengl backend :P will add one to the opengl backend as well! 
Btw shall we rebase this on top of master and enable aot test on metal backend as well? https://github.com/taichi-dev/taichi/blob/master/tests/python/test_aot.py#L31

File: taichi/backends/wasm/aot_module_builder_impl.h (hunk 1 above)
Comment (k-ye) [reply taichi/backends/wasm/aot_module_builder_impl.h:27]: It kind of is. Users will be able to retrieve the info of the field (represented by this SNode) via the identifier.

**Pull request merged**